compiler/zrc_typeck/src/typeck/ty.rs

```diff
@@ -180,4 +337,203 @@ mod tests {
+                                ParserType(spanned!(
+                                    27,
+                                    ParserTypeKind::Ptr(Box::new(ParserType(spanned!(
+                                        28,
+                                        ParserTypeKind::Identifier("Node"),
+                                        32
+                                    )))),
+                                    32
+                                ))
+                            ),
+                            32
+                        )
+                    ],
+                    34
+                ))),
+                34
+            )),
+            "Node",
+        );
+
+        assert!(result.is_ok());
+        let resolved_ty = result.expect("should resolve successfully");
+        if let TastType::Struct(fields) = resolved_ty {
+            assert_eq!(fields.len(), 2);
+            assert_eq!(fields.get("value"), Some(&TastType::I32));
+            // The pointer to self should be replaced with pointer to empty struct
+            assert_eq!(
+                fields.get("next"),
+                Some(&TastType::Ptr(Box::new(TastType::unit())))
+            );
+        } else {
+            panic!("Expected struct type");
+        }
+    }
+
+    #[test]
+    fn self_referential_type_not_behind_pointer_produces_error() {
+        // struct { value: i32, next: Node }
+        // where Node is the name being defined (ERROR: not behind pointer)
+        let result = resolve_type_with_self_reference(
+            &TypeCtx::default(),
+            ParserType(spanned!(
+                0,
+                ParserTypeKind::Struct(KeyTypeMapping(spanned!(
+                    7,
+                    vec![
+                        spanned!(
+                            9,
+                            (
+                                spanned!(9, "value", 14),
+                                ParserType(spanned!(16, ParserTypeKind::Identifier("i32"), 19))
+                            ),
+                            19
+                        ),
+                        spanned!(
+                            21,
+                            (
+                                spanned!(21, "next", 25),
+                                ParserType(spanned!(27, ParserTypeKind::Identifier("Node"), 31))
+                            ),
+                            31
+                        )
+                    ],
+                    33
+                ))),
+                33
+            )),
+            "Node",
+        );
+
+        assert!(result.is_err());
+        let err = result.expect_err("should produce error");
+        assert_eq!(
+            err,
+            Diagnostic(
+                Severity::Error,
+                spanned!(
+                    0,
+                    DiagnosticKind::SelfReferentialTypeNotBehindPointer("Node".to_string()),
+                    33
```

Comment:
Why is this span 0..33? Shouldn't the thing erroring be just the "x: Node" line?